akka-typed/src/main/scala/akka/typed/internal/SystemMessage.scala

```diff
@@ -0,0 +1,230 @@
+  import SystemMessageList._
+
+  /**
+   * Indicates if the list is empty or not. This operation has constant cost.
+   */
+  final def isEmpty: Boolean = head eq null
+
+  /**
+   * Indicates if the list has at least one element or not. This operation has constant cost.
+   */
+  final def nonEmpty: Boolean = head ne null
+
+  /**
+   * Indicates if the list is empty or not. This operation has constant cost.
+   */
+  final def size: Int = sizeInner(head, 0)
+
+  /**
+   * Gives back the list containing all the elements except the first. This operation has constant cost.
+   *
+   * *Warning:* as the underlying list nodes (the [[SystemMessage]] instances) are mutable, care
+   * should be taken when passing the tail to other methods. [[SystemMessage#unlink]] should be
+   * called on the head if one wants to detach the tail permanently.
+   */
+  final def tail: EarliestFirstSystemMessageList = new EarliestFirstSystemMessageList(head.next)
+
+  /**
+   * Reverses the list. This operation mutates the underlying list. The cost of the call to reverse is linear in the
+   * number of elements.
+   *
+   * The type of the returned list is of the opposite order: [[LatestFirstSystemMessageList]]
+   */
+  final def reverse: LatestFirstSystemMessageList = new LatestFirstSystemMessageList(reverseInner(head, null))
+
+  /**
+   * Attaches a message to the current head of the list. This operation has constant cost.
+   */
+  final def ::(msg: SystemMessage): EarliestFirstSystemMessageList = {
+    assert(msg ne null)
+    msg.next = head
+    new EarliestFirstSystemMessageList(msg)
+  }
+
+  /**
+   * Prepends a list in a reversed order to the head of this list. The prepended list will be reversed during the process.
+   *
+   * Example: (3, 4, 5) reversePrepend (2, 1, 0) == (0, 1, 2, 3, 4, 5)
+   *
+   * The cost of this operation is linear in the size of the list that is to be prepended.
+   */
+  final def reverse_:::(other: LatestFirstSystemMessageList): EarliestFirstSystemMessageList = {
+    var remaining = other
+    var result = this
+    while (remaining.nonEmpty) {
+      val msg = remaining.head
+      remaining = remaining.tail
+      result ::= msg
+    }
+    result
+  }
+
+}
+
+/**
+ * System messages are handled specially: they form their own queue within
+ * each actor’s mailbox. This queue is encoded in the messages themselves to
+ * avoid extra allocations and overhead. The next pointer is a normal var, and
+ * it does not need to be volatile because in the enqueuing method its update
+ * is immediately succeeded by a volatile write and all reads happen after the
+ * volatile read in the dequeuing thread. Afterwards, the obtained list of
+ * system messages is handled in a single thread only and not ever passed around,
+ * hence no further synchronization is needed.
+ *
+ * INTERNAL API
+ *
+ * <b>NEVER SEND THE SAME SYSTEM MESSAGE OBJECT TO TWO ACTORS</b>
+ */
+private[typed] sealed trait SystemMessage extends Serializable {
```

Comment:
Yes, that is true. Are there alternative serializers that use the same marker interface for their purposes?